### PR TITLE
formulae: add livecheck throttle, throttled_formulae: removals

### PR DESCRIPTION
**Pull request by cho-m**
- [ ] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [ ] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [ ] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

```
Matched Versions:
0.10.0, 0.10.15, 0.11.0, 0.12.0, 0.12.15, 0.13.15, 1.0.105, 1.0.120, 1.0.135, 1.0.15, 1.0.150, 1.0.30, 1.0.45, 1.0.60, 1.0.75, 1.0.90, 1.1.15, 1.1.30, 1.1.45, 1.1.60, 1.10.0, 1.10.15, 1.10.30, 1.10.45, 1.11.0, 1.11.105, 1.11.120, 1.11.135, 1.11.15, 1.11.150, 1.11.165, 1.11.180, 1.11.195, 1.11.210, 1.11.225, 1.11.240, 1.11.255, 1.11.270, 1.11.285, 1.11.30, 1.11.45, 1.11.60, 1.11.75, 1.11.90, 1.2.0, 1.2.15, 1.2.30, 1.3.0, 1.3.15, 1.3.30, 1.3.45, 1.4.0, 1.4.15, 1.4.30, 1.4.45, 1.4.60, 1.4.75, 1.4.90, 1.5.0, 1.5.15, 1.6.0, 1.6.15, 1.6.30, 1.6.45, 1.7.0, 1.7.105, 1.7.120, 1.7.135, 1.7.15, 1.7.150, 1.7.165, 1.7.180, 1.7.195, 1.7.210, 1.7.225, 1.7.240, 1.7.255, 1.7.270, 1.7.285, 1.7.30, 1.7.300, 1.7.315, 1.7.330, 1.7.345, 1.7.360, 1.7.45, 1.7.60, 1.7.75, 1.7.90, 1.8.0, 1.8.105, 1.8.120, 1.8.135, 1.8.15, 1.8.150, 1.8.165, 1.8.180, 1.8.30, 1.8.45, 1.8.60, 1.8.75, 1.8.90, 1.9.0, 1.9.105, 1.9.120, 1.9.135, 1.9.15, 1.9.150, 1.9.165, 1.9.180, 1.9.195, 1.9.210, 1.9.225, 1.9.240, 1.9.255, 1.9.270, 1.9.285, 1.9.30, 1.9.300, 1.9.315, 1.9.330, 1.9.345, 1.9.360, 1.9.375, 1.9.45, 1.9.60, 1.9.75, 1.9.90

aws-sdk-cpp: 1.11.285 ==> 1.11.285
```

### CODE REVIEW

**Comment by cho-m** [comment]
> @cho-m Question, does it mean including a throttling at the livecheck level makes the whole throttling file as it currently exist pointless, once those will be done?

Essentially, if the formulae are on autobump. Otherwise, throttling file still prevents `brew bump-formula-pr --version ...`

The main reason for this change is to help autobump since missing a version means we need to wait for next throttled version.

In future, it may be worth adding some DSL for readability (rather than comments, audit files, and regex modifications), e.g.
```rb
livecheck do
  url ...
  regex(/.../)
  throttle 10
end
```

---

> Also, I guess bot shouldn't set [ long build](https://github.com/Homebrew/homebrew-core/labels/long%20build) Needs CI-long-timeout if it finds [ CI-syntax-only](https://github.com/Homebrew/homebrew-core/labels/CI-syntax-only) Change only affects brew syntax, not the install. Only run syntax CI. , should it?

The extra labels don't block run so behavior is fine but not ideal. Current limitation is the corresponding JavaScript action code doesn't provide a way to create dependent labels (e.g. if `CI-syntax-only` then don't label `long build`).

**Comment by SMillerDev** [comment]
I thought in general we didn't throttle livecheck since then it isn't really a livecheck anymore.

CC @samford ?

**Comment by cho-m** [comment]
> I thought in general we didn't throttle livecheck since then it isn't really a livecheck anymore.
> 
> CC @samford ?

The problem is autobump. It has no idea about throttle and can only attempt to create PRs on whatever livecheck outputs.

There is a chance it never updates a formula due to this, e.g. 
- Let's say `aws-sdk-cpp` releases `1.11.300` and `1.11.301`
- Bot tries autobump, sees `1.11.301` and hits audit failure
- Bot will keep trying for `1.11.302`, and onward
- Let's say we get really unlucky and they release `1.11.315` and then `1.11.316` on same day and we miss it again.

Our throttle + autobump has pretty much become a block updates feature.

**Comment by cho-m** [comment]
> > The problem is autobump. It has no idea about throttle and can only attempt to create PRs on whatever livecheck outputs.
> 
> Is there any reason why `brew bump` (and/or the `bump-packages` action) can't be updated to respect `throttled_formulae.json`?

`brew bump` just forwards data from `livecheck` to `brew bump-formula-pr` and has no versioning awareness. Moving the versioning info into `bump` will result in duplicating more `livecheck` logic as we will need to get all versions and scan through them in similar fashion rather than reusing existing logic.

Based on general direction towards automated updates and easier maintainability, I feel that `livecheck` itself should be throttle-aware and be able to present the throttled version perhaps along-side the latest version.

---

For example, a self-documenting DSL like:
```rb
  livecheck do
    throttle 15
  end
```

That can be coupled with code like something I quickly hacked up in https://github.com/Homebrew/brew/pull/16918

Can allow us to get data like:
```console
❯ brew bump aws-sdk-cpp
==> aws-sdk-cpp is up to date!
Current formula version:  1.11.285
Latest livecheck version: 1.11.285 (throttled)
Latest Repology version:  1.11.285
Open pull requests:       aws-sdk-cpp: throttle livecheck output (https://github.com/Homebrew/homebrew-core/pull/166501)
Closed pull requests:     none

❯ brew livecheck aws-sdk-cpp
aws-sdk-cpp: 1.11.285 ==> 1.11.288
```

Without having to push version calculation logic into `brew bump`.

**Comment by cho-m** [comment]
> (like [`bump-formula-pr`](https://github.com/Homebrew/brew/blob/e3797d39bde3fe4c4b9c5c3b17dfd653dc592dc9/Library/Homebrew/dev-cmd/bump-formula-pr.rb#L473))?

That's part of the problem. We don't want that behavior as it causes a race condition.

**Review by carlocab — COMMENTED** [review]
It seems like the solution is to make `brew bump[-formula-pr]` aware of the latest throttled version, which should be handled in `brew`, and does seem to be most sensibly handled by `livecheck`.

File: Formula/c/checkov.rb [hunk 1]
```diff
@@ -7,6 +7,12 @@ class Checkov < Formula
   sha256 "dca00fe8f0e04fd4fde88de5143bbe01aae7966351cd0f70f69456313bd43428"
   license "Apache-2.0"
 
+  livecheck do
+    url "https://pypi.org/rss/project/checkov/releases.xml"
+    regex(%r{<title>v?(\d+(?:\.\d+)+)</title>}i)
```
Comment (cho-m) [inline Formula/c/checkov.rb:12]: Or git as I mentioned in #166497
```suggestion
    url "https://github.com/bridgecrewio/checkov.git"
    regex(/^v?(\d+(?:\.\d+)+)$/i)
```

`git` is probably safer in avoiding risk of multiple pre-releases (alpha/beta/rc) that can fill up RSS feed, though `checkov` never did pre-releases before (even going from v2 to v3).

PyPI RSS is currently faster as less content. Running `git ls-remote --tags` is 199KB (and will continue to increase) while RSS xml is 11KB (should remain relatively same as there is a limit). 

Another option is GitHub API, but we probably need to be more careful due to autobump being at risk of hitting rate limits.

**Comment by samford** [comment]
I've opened https://github.com/Homebrew/brew/pull/16932 to update the `LivecheckUrlProvided` RuboCop to allow `livecheck` blocks that only include a `#throttle` call (it was a bit trickier, since `#throttle` can appear alongside other livecheck DSL methods). Once that's sorted, we can switch `aws-sdk-cpp` back to only using `#throttle`.

**Review by carlocab — APPROVED** [review]
LGTM, but I think we need a new `brew` tag first? (Tho this probably doesn't affect users on a `brew` tag if they're on the JSON API)

**Comment by samford** [comment]
> LGTM, but I think we need a new `brew` tag first? (Tho this probably doesn't affect users on a brew tag if they're on the JSON API)

Yep, this involves changes to the `livecheck` block DSL, so it will need a `brew` release before we can use `#throttle` in formulae. If we merge this before that point, it will throw an `undefined method` error whenever the formula is loaded.

**Comment by cho-m** [comment]
> I've opened [Homebrew/brew#16932](https://github.com/Homebrew/brew/pull/16932) to update the `LivecheckUrlProvided` RuboCop to allow `livecheck` blocks that only include a `#throttle` call (it was a bit trickier than `#skip`, `#formula`, etc., since `#throttle` can appear alongside other livecheck DSL methods). Once that's sorted, we can switch `aws-sdk-cpp` back to only using `#throttle`.

Thanks. Will update this PR afterward.

---

> LGTM, but I think we need a new `brew` tag first? (Tho this probably doesn't affect users on a `brew` tag if they're on the JSON API)

Yes. This PR should remain a draft until then to avoid breaking non-API users on stable `brew`.

File: Formula/a/aws-sdk-cpp.rb [hunk 2]
```diff
@@ -1,13 +1,17 @@
...
   homepage "https://github.com/aws/aws-sdk-cpp"
-  # aws-sdk-cpp should only be updated every 15 releases on multiples of 15
   url "https://github.com/aws/aws-sdk-cpp.git",
       tag:      "1.11.285",
       revision: "8213a588808acf5148b0f739e0b558879955310f"
   license "Apache-2.0"
   head "https://github.com/aws/aws-sdk-cpp.git", branch: "main"
 
+  livecheck do
+    url :stable
```
Comment (samford) [inline Formula/a/aws-sdk-cpp.rb:11]:
```suggestion
```

https://github.com/Homebrew/brew/pull/16932 is merged, so we can omit `url` here now.

File: Formula/c/checkov.rb (hunk 1 above)
Comment (samford) [inline Formula/c/checkov.rb:12]:
```suggestion
    regex(/^v?(\d+(?:\.\d+)+)$/i)
    strategy :xml do |xml, regex|
      xml.get_elements("//item/title").map { |item| item.text[regex, 1] }
    end
```

Checking the PyPI RSS feed makes sense to me, though we should use the `Xml` strategy to make sure we're only checking `item` titles.

**Review by samford — APPROVED** [review]

**Comment by samford** [comment]
Looks good to me. Should be good to merge after the `awscli@1` conflict is resolved.

**Review by chenrui333 — APPROVED** [review]

**Pull request merged**